fleetconfig-controller/internal/controller/addon.go

```diff
@@ -0,0 +1,218 @@
+			// pass URL directly
+			args = append(args, fmt.Sprintf("--filename=%s", manifests))
+		case manifestTypeRaw:
+			filename, cleanup, err := file.TmpFile([]byte(manifests), "yaml")
+			if cleanup != nil {
+				defer cleanup()
+			}
+			if err != nil {
+				return err
+			}
+			args = append(args, fmt.Sprintf("--filename=%s", filename))
+		}
+
+		if a.HubRegistration {
+			args = append(args, "--hub-registration")
+		}
+		if a.Overwrite {
+			args = append(args, "--overwrite")
+		}
+		if a.ClusterRoleBinding != "" {
+			args = append(args, fmt.Sprintf("--cluster-role-bind=%s", a.ClusterRoleBinding))
+		}
+
+		cmd := exec.Command(clusteradm, args...)
+		out, err := exec_utils.CmdWithLogs(ctx, cmd, "waiting for 'clusteradm addon create' to complete...")
+		if err != nil {
+			return fmt.Errorf("failed to create addon: %v, output: %s", err, string(out))
+		}
+	}
+	return nil
+}
+
+func checkManifestType(manifests string) manifestType {
+	if strings.HasPrefix(manifests, "http") || strings.HasPrefix(manifests, "https") {
+		return manifestTypeHTTP
+	}
+	return manifestTypeRaw
+}
+
+func handleAddonDelete(ctx context.Context, addonC *addonapi.Clientset, fc *v1alpha1.FleetConfig, addons []string) error {
+	logger := log.FromContext(ctx)
+	logger.V(0).Info("deleteAddOns", "fleetconfig", fc.Name)
+
+	for _, addonName := range addons {
+		// get the addon template, so we can extract spec.addonName
+		addon, err := addonC.AddonV1alpha1().AddOnTemplates().Get(ctx, addonName, metav1.GetOptions{})
+		if err != nil && !kerrs.IsNotFound(err) {
+			return fmt.Errorf("failed to delete addon %s: %v", addonName, err)
+		}
+
+		// delete the addon template
+		if addon != nil {
+			err = addonC.AddonV1alpha1().AddOnTemplates().Delete(ctx, addonName, metav1.DeleteOptions{})
+			if err != nil && !kerrs.IsNotFound(err) {
+				return fmt.Errorf("failed to delete addon %s: %v", addonName, err)
+			}
+		}
+
+		// get the addon name without a version suffix
+		unversionedName := addon.Spec.AddonName
+
+		// check if there are any remaining addon templates for the same addon (different versions of the same addon)
+		allAddons, err := addonC.AddonV1alpha1().AddOnTemplates().List(ctx, metav1.ListOptions{})
+		if err != nil && !kerrs.IsNotFound(err) {
+			return fmt.Errorf("failed to delete addon %s: %v", addonName, err)
+		}
+
+		remainingAddons := []addonv1alpha1.AddOnTemplate{}
+		for _, a := range allAddons.Items {
+			if a.Spec.AddonName == unversionedName {
+				remainingAddons = append(remainingAddons, a)
+			}
+		}
+		// if any remain, nothing else to do
+		if len(remainingAddons) > 0 {
+			logger.V(0).Info("deleted addon", "AddOnTemplate", addonName)
+			continue
+		}
+
+		// if none remain, delete the ClusterManagementAddOn
+		err = addonC.AddonV1alpha1().ClusterManagementAddOns().Delete(ctx, unversionedName, metav1.DeleteOptions{})
+		if err != nil && !kerrs.IsNotFound(err) {
+			return fmt.Errorf("failed to delete addon %s: %v", addonName, err)
+		}
+
```

Comment:
Should this be pulled out of the range loop?

IE, you use the range loop just to delete all the addons that are passed in, and once you finish deleting them, you check if there are any remaining. If none remain, delete the clusterManagementAddon.

That way you also wont need multiple spots to log `deleted addon`. You can log that once at the end of each range loop, and you can log that you deleted the ClusterManagementAddOn once only if there were no addOns remaining.